rfd/0136-modern-signature-algorithms.md

```diff
@@ -0,0 +1,881 @@
+#### Cloud
+
+Cloud will be able to select their preferred defaults by configuring them in the
+`teleport.yaml`.
+Cloud users will be able to change the CA algorithms by modifying the
+`cluster_auth_preference` and performing a CA rotation.
+
+### Subjects (users/hosts/DBs/JWTs)
+
+It will be great to update the CA key algorithms for security and performance
+benefits, but what users really see on a day-to-day basis is their user keys.
+
+"Subjects" that have certificates issued by the Teleport CAs include:
+
+* Teleport users via `tsh login`
+* Teleport users via `tsh app login`
+* Teleport users via `tsh db login`
+* Teleport users via Teleport Connect
+* Teleport services (ssh, app, db, kube, windows desktop, etc)
+* Machine ID (`tbot`)
+* Teleport Plugins
+* OpenSSH hosts
+* Databases
+* Proxies communicating with remote database services
+
+All of these currently generate an RSA2048 keypair locally, send the public key
+to the auth server, and receive signed certificates of some variety.
+
+The `allowed_subject_algorithms` field described in the CA configuration section
+will control which key algorithms can be used by subject keys.
+These should be in preference order, compliant clients should pick the first
+algorithm from the list that they support.
+This will allow us to introduce new algorithms to the list without breaking
+compatibility with existing clients.
+
+In some cases, Teleport services may choose a preferred algorithm from the list
+that is not the first one they support.
+For example, the Database CA signs DB certs that are often fairly long-lived, as
+well as short-lived certs used by proxies connecting to remote database services
+on behalf of the user, it may be preferable to select different algorithms for
+each case.
+
+### Splitting user SSH and TLS private keys
+
+In order to support the use of different key types for SSH and TLS, users will
+need to start generating two different private keys, one for each protocol.
+This will change the disk layout of the ~/.tsh directory:
+
+```diff
+  ~/.tsh/                             --> default base directory
+  ├── current-profile                 --> file containing the name of the currently active profile
+  ├── one.example.com.yaml            --> file containing profile details for proxy "one.example.com"
+  ├── two.example.com.yaml            --> file containing profile details for proxy "two.example.com"
+  ├── known_hosts                     --> trusted certificate authorities (their keys) in a format similar to known_hosts
+  └── keys                            --> session keys directory
+     ├── one.example.com              --> Proxy hostname
+     │   ├── certs.pem                --> TLS CA certs for the Teleport CA
+-    │   ├── foo                      --> Private Key for user "foo"
+-    │   ├── foo.pub                  --> Public Key
++    │   ├── foo                      --> SSH Private Key for user "foo"
++    │   ├── foo.pub                  --> SSH Public Key
```

Comment:
I'd pick names that don't clash with the previous filenames - using the wrong private key with an X.509 cert is going to lead to really strange errors, I bet.

Reply:
I wanted to avoid breaking all the openssh configs out there that read the priv/pub from this location, and figured not much actually reads an x509 key/cert from here (except tsh which is the one writing this and should know the right location). Arguably we should change the x509 cert location, but I think it would be okay to leave the ssh keypair where it is

Reply:
In the latest revision I change the name of all TLS certs, but leave the SSH key/cert where it is